### PR TITLE
Add BDD tests for functions

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## Release notes: product changes
Add basic BDD tests for functions & implement validation.

## Implementation
* Adds BDD tests for functions
* Adds validation of function definitions & calls.
* Adds stratification validation.
* Cleaner plumbing of function annotations

### CODE REVIEW

File: compiler/annotation/mod.rs [hunk 1]
```diff
@@ -78,6 +78,18 @@ typedb_error!(
             "Error analysing return reduction.",
             ( typedb_source : Box<AnnotationError> )
         ),
+        CouldNotResolveReturnType(
```
Comment (flyingsilverfin) [inline compiler/annotation/mod.rs:81]: You can definitely reorder them if you want? But error 2 is different (to do with call arguments)?

File: compiler/annotation/function.rs [hunk 2]
```diff
@@ -48,10 +54,18 @@ pub struct AnnotatedFunction {
     pub variable_registry: VariableRegistry,
     pub parameter_registry: ParameterRegistry,
     pub arguments: Vec<Variable>,
+    pub argument_annotations: Vec<FunctionParameterAnnotation>,
     pub stages: Vec<AnnotatedStage>,
     pub return_: AnnotatedFunctionReturn,
 }
 
+impl AnnotatedFunction {
+    pub(crate) fn get_annotated_signature(&self) -> AnnotatedFunctionSignature {
```
Comment (flyingsilverfin) [inline compiler/annotation/function.rs:63]: so i general, im not a huge fan of cloning things as they are returned - better to use Refs (Cows/maybeOwned) and clone on the receiver end when you need it to be owned

Comment (flyingsilverfin) [reply compiler/annotation/function.rs:63]: Or store an AnnotatedFunctionSignature and pass an & back, then clone that if needed

File: function/function_manager.rs [hunk 3]
```diff
@@ -110,27 +124,73 @@ impl FunctionManager {
...
         Ok(functions)
     }
 
+    pub fn undefine_function(&self, snapshot: &mut impl WritableSnapshot, name: &str) -> Result<(), FunctionError> {
+        let definition_key = match self.get_function_key(snapshot, name) {
+            Err(source) => Err(FunctionError::FunctionRetrieval { source }),
+            Ok(None) => Err(FunctionError::FunctionNotFound {}),
+            Ok(Some(key)) => Ok(key),
+        }?;
+        snapshot.delete(definition_key.into_storage_key().into_owned_array());
```
Comment (flyingsilverfin) [inline function/function_manager.rs:145]: I you need a validation here to make sure it's not used in another function?

File: function/lib.rs [hunk 4]
```diff
@@ -23,5 +23,6 @@ typedb_error!(
         CreateFunctionEncoding(6, "Encoding error while trying to create function.", ( source: EncodingError )),
         FunctionRetrieval(7, "Error retrieving function.", (  source: FunctionReadError )),
         CommittedFunctionParseError(8, "Error while parsing committed.", ( typedb_source: typeql::Error )),
+        StratificationViolation(9, "StratificationViolation."), // TODO: Message
```
Comment (flyingsilverfin) [inline function/lib.rs:26]: improve please, this is not useful to users

File: ir/translation/function.rs [hunk 5]
```diff
@@ -59,15 +54,39 @@ pub fn translate_function_from(
...
+    for arg in &signature.args {
         let var = context.get_variable(arg.var.name().unwrap()).ok_or_else(|| {
             FunctionRepresentationError::FunctionArgumentUnused {
                 argument_variable: arg.var.name().unwrap().to_owned(),
                 declaration: declaration.unwrap().clone(),
             }
         })?;
     }
-    Ok(Function::new(name, context, value_parameters, arguments, Some(argument_labels), Some(output.clone()), body))
+    // Check return declaration aligns with definition
```
Comment (flyingsilverfin) [inline ir/translation/function.rs:65]: nice

File: ir/translation/function.rs [hunk 6]
```diff
@@ -84,6 +103,22 @@ pub(crate) fn translate_function_block(
                 typedb_source: err,
             })?;
 
+    let mut illegal_stages = stages.iter().filter(|stage| {
+        match stage {
+            TranslatedStage::Insert { .. } | TranslatedStage::Delete { .. } => true,
+            TranslatedStage::Require(_) => false, // Is require allowed
```
Comment (flyingsilverfin) [inline ir/translation/function.rs:109]: yeah i think it can be allowed!

Comment (flyingsilverfin) [reply ir/translation/function.rs:109]: it's just a filter for eliminating answers with empty variables

File: query/define.rs [hunk 7]
```diff
@@ -126,11 +128,15 @@ fn process_type_definitions(
...
 fn process_function_definitions(
     snapshot: &mut impl WritableSnapshot,
-    type_manager: &TypeManager,
+    function_manager: &FunctionManager,
     definables: &[Definable],
 ) -> Result<(), DefineError> {
-    filter_variants!(Definable::Function : definables)
-        .try_for_each(|function| define_function(snapshot, type_manager, function))?;
+    let functions = filter_variants!(Definable::Function: definables);
+    if functions.clone().next().is_some() {
```
Comment (flyingsilverfin) [inline query/define.rs:135]: why clone?

File: query/query_manager.rs [hunk 8]
```diff
@@ -72,6 +82,10 @@ impl QueryManager {
         let arced_premable = Arc::new(translated_preamble);
         let arced_stages = Arc::new(translated_stages);
         let arced_fetch = Arc::new(translated_fetch);
+        match validate_no_cycles(&arced_premable.iter().enumerate().map(|(i, translated)| (i, translated)).collect()) {
+            Ok(_) => {}
+            Err(typedb_source) => return Err(QueryError::FunctionRetrieval { typedb_source }),
```
Comment (flyingsilverfin) [inline query/query_manager.rs:87]: No, we can generalise this error now?

Comment (flyingsilverfin) [reply query/query_manager.rs:87]: The name of it at least

File: tests/behaviour/debug/debug.feature [hunk 9]
```diff
@@ -5,4 +5,58 @@
 Feature: Debugging Space
 
   # Paste any scenarios below for debugging.
-  # Do not commit any changes to this file.
\ No newline at end of file
+  # Do not commit any changes to this file.
+
+  Background: Set up database
```
Comment (flyingsilverfin) [inline tests/behaviour/debug/debug.feature:10]: clean up

**Review by flyingsilverfin — CHANGES_REQUESTED** [review]

File: function/function_manager.rs (hunk 3 above)
Comment (krishnangovindraj) [reply function/function_manager.rs:145]: Too much work to build out the infra for that if I need to switch to performance.
I have a BDD that makes sure the commit fails.

File: compiler/annotation/function.rs (hunk 2 above)
Comment (krishnangovindraj) [reply compiler/annotation/function.rs]: Reworked the whole thing. Please re-review the structure of `AnnotatedFunction`

File: query/query_manager.rs (hunk 8 above)
Comment (krishnangovindraj) [reply query/query_manager.rs]: Updated. Please check & see if we need to box the error. I don't know what the current strategy is. Box everywhere?

**Review by krishnangovindraj — COMMENTED** [review]
responses

File: compiler/annotation/pipeline.rs [hunk 10]
```diff
@@ -222,8 +228,7 @@ fn annotate_stage(
     parameters: &ParameterRegistry,
     snapshot: &impl ReadableSnapshot,
     type_manager: &TypeManager,
-    schema_function_annotations: &IndexedAnnotatedFunctions,
-    preamble_function_annotations: Option<&AnnotatedUnindexedFunctions>,
+    annotated_function_signatures: &dyn AnnotatedFunctionSignatures,
```
Comment (flyingsilverfin) [inline compiler/annotation/pipeline.rs:231]: This trait dyn thing works for me!

File: function/function_manager.rs (hunk 3 above)
Comment (flyingsilverfin) [reply function/function_manager.rs:145]: leave a TODO then please

**Review by flyingsilverfin — APPROVED** [review]

File: query/query_manager.rs (hunk 8 above)
Comment (flyingsilverfin) [reply query/query_manager.rs]: Right now we only box the results that have Large errors in them - if Clippy doesn't warn us on that we don't need to atm.

File: function/function_manager.rs (hunk 3 above)
Comment (krishnangovindraj) [reply function/function_manager.rs:145]: It's already there.

**Pull request merged**